File: branches/LOFAR-Task7336/LCS/MessageBus/src/MsgBus.cc
#include "lofar_config.h"

#include <MessageBus/MsgBus.h>
#include <Common/LofarLogger.h>

#ifdef HAVE_QPID
#include <qpid/types/Exception.h>
#include <qpid/messaging/exceptions.h>
//#include <qpid/messaging/Logger.h>

using namespace qpid::messaging;

namespace LOFAR {

  static Duration TimeOutDuration(double secs)
  {
    if (secs > 0.0)
      return (Duration)(1000.0 * secs);

    return Duration::FOREVER;
  }

  FromBus::FromBus(const std::string &address, const std::string &options, const std::string &broker)
  try:
    itsConnection(broker),
    itsNrMissingACKs(0)
  {
    itsConnection.open();
	cout << "Connected to: " << itsConnection.getUrl() << endl;

    itsSession = itsConnection.createSession();

    Address addr(address+options);
    Receiver receiver = itsSession.createReceiver(addr);
    receiver.setCapacity(1);
	cout << "Receiver started at queue: " << receiver.getName() << endl;
  } catch(const qpid::types::Exception &ex) {
    THROW(MessageBusException, ex.what());
  }

 
  FromBus::~FromBus(void)
  {
    if (itsNrMissingACKs) {
//      LOG_ERROR_STR("Queue " << itsQueueName << " on broker " << itsBrokerName << " has " << itsNrMissingACKs << " messages not ACK'ed ");
	}

    try {
      // Make sure all requests are finished
      itsConnection.close();
    } catch(const qpid::types::Exception &ex) {
      LOG_FATAL_STR("Exception in destructor, cannot guarantee message delivery: " << ex.what());
    }
  }

  bool FromBus::getMessage(LOFAR::Message &msg, double timeout) // timeout 0.0 means blocking
  {
    Receiver next;
	qpid::messaging::Message	qmsg;
	cout << "waiting for message..." << endl;
    if (itsSession.nextReceiver(next,TimeOutDuration(timeout))) {
		cout << "message available on queue: " << next.getName() << endl;
        itsNrMissingACKs++;
        if (next.get(qmsg)) {
			msg = LOFAR::Message(qmsg);
			return true;
		}
    }
    return false;
  }

  void FromBus::ack(LOFAR::Message &msg)
  {
     itsSession.acknowledge(msg.qpidMsg());
     itsNrMissingACKs --;
  }

  void FromBus::nack(LOFAR::Message &msg)
  {
     itsSession.release(msg.qpidMsg());
     itsNrMissingACKs--;
  }

  void FromBus::reject(LOFAR::Message &msg)
  {
     itsSession.reject(msg.qpidMsg());
     itsNrMissingACKs --;
  }

  bool FromBus::addQueue(const std::string &address, const std::string &options)
  {
     try
     {
        Address addr(address+options);
        Receiver receiver = itsSession.createReceiver(addr);
        receiver.setCapacity(1);
		cout << "Receiver started at queue: " << receiver.getName() << endl;
     } catch(const qpid::types::Exception &ex) {
       //THROW(MessageBusException, ex.what());
       return false;
     }
     return true;
  }

  ToBus::ToBus(const std::string &address, const std::string &options, const std::string &broker) 
  try:
    itsConnection(broker)
  {
     itsConnection.open();
	 cout << "Connected to: " << itsConnection.getUrl() << endl;

     itsSession = itsConnection.createSession();
     Address addr(address+options);
     itsSender = itsSession.createSender(addr);
 	 cout << "Sender created: " << itsSender.getName() << endl;
  } catch(const qpid::types::Exception &ex) {
    THROW(MessageBusException, ex.what());
  }

  ToBus::~ToBus(void)
  {

    try {
      // Make sure all requests are finished
      istSender.close();
      itsConnection.close();
    } catch(const qpid::types::Exception &ex) {
      LOG_FATAL_STR("Exception in destructor, cannot guarantee message delivery: " << ex.what());
    }
  }

  void ToBus::send(const std::string &msg)
  {
    LOFAR::Message tosend(msg);
    itsSender.send(tosend.qpidMsg(), true);
  }

  void ToBus::send(LOFAR::Message& msg)
  {
    itsSender.send(msg.qpidMsg(), true);
  }

} // namespace LOFAR

#endif

File: branches/LOFAR-Task7336/LCS/MessageBus/src/msgbus.py
#!/usr/bin/python
# Copyright (C) 2012-2013  ASTRON (Netherlands Institute for Radio Astronomy)
# P.O. Box 2, 7990 AA Dwingeloo, The Netherlands
#
# This file is part of the LOFAR software suite.
# The LOFAR software suite is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# The LOFAR software suite is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
#
# id.. TDB

try:
  import qpid.messaging
  enabled = True
except ImportError:
  enabled = False

import os
import lofar.messagebus.message as message

# Candidate for a config file
broker="127.0.0.1" 
options="create:never"

class BusException(Exception):
    pass

class Session:
    def __init__(self, broker):
        self.connection = qpid.messaging.Connection(broker)
        self.connection.reconnect = True

        try:
            self.connection.open()
            self.session = self.connection.session() 
        except qpid.messaging.MessagingError, m:
            raise BusException(m)

    def __del__(self):
        self.connection.close()

    def address(self, queue, options):
        return "%s%s; {%s}" % (self._queue_prefix(), queue, options)

    def _queue_prefix(self):
      return os.environ.get("QUEUE_PREFIX", "")

class ToBus(Session):
    def __init__(self, queue, options=options, broker=broker):
        Session.__init__(self, broker)

        try:
            self.sender = self.session.sender(self.address(queue, options))
        except qpid.messaging.MessagingError, m:
            raise BusException(m)

    def __del__(self):
        self.connection.close()

    def send(self, msg, verbatim=False):
        # (Re)generate the QPID message body
        if not verbatim:
            msg.generate_message()

        try:
            self.sender.send(msg.qpidMsg)
        except qpid.messaging.SessionError, m:
            raise BusException(m)

class FromBus(Session):
    def __init__(self, queue, options=options, broker=broker):
        Session.__init__(self, broker)

        self.add_queue(queue, options)

    def add_queue(self, queue, options=options):
        try:
            receiver = self.session.receiver(self.address(queue, options))

            # Need capacity >=1 for 'self.session.next_receiver' to function across multiple queues
            receiver.capacity = 1 #32
        except qpid.messaging.MessagingError, m:
            raise BusException(m)

    def get(self, timeout=None):
        msg = None

        try:
            receiver = self.session.next_receiver(timeout)
            if receiver != None: msg = receiver.fetch() # receiver.get() is better, but requires qpid 0.31+
        except qpid.messaging.exceptions.Empty, e:
            return None

        if msg is None:
          return None
        else:
          return message.Message(qpidMsg=msg)

    def ack(self, msg):
        self.session.acknowledge(msg.qpidMsg)

